### PR TITLE
Lesson 5-7: Driver for character device

### PR DESCRIPTION
**Pull request by D-Alexandr**
Added Makefile and driver for character device.

@oleg-h2o , @galkinletter please check my homework

### CODE REVIEW

File: 02-KernelModules_CharacterDevice/chrdev_driver.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+	cdev = cdev_alloc();
+ 	
+	 if (!cdev) {
+		result = -ENOMEM;
+		goto fail; 
+	}
+	cdev_init(cdev, &chrdev_fops);
+	cdev->owner   = THIS_MODULE;
+	if (cdev_add(cdev, dev, chrdev_nr_devs)) {
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/chrdev_driver.c]:
```suggestion
	if (cdev_add(cdev, dev, chrdev_nr_devs)) { 
			printk(KERN_WARNING "chrdev: cdev_add error \n");
		}
		printk(KERN_INFO "chrdev: %d:%d \n", chrdev_major, chrdev_minor);
	
```

printk(KERN_WARNING  "chrdev: cdev_add error \n");

Please do it error handling  for this.

Think of fatal errors and their handling.

**Review by galkinletter — CHANGES_REQUESTED** [review]
Please do it error handling for some events

Think of fatal errors and their handling.

**Comment by D-Alexandr** [comment]
Fixed. @galkinletter  please check

File: 02-KernelModules_CharacterDevice/chrdev_driver.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+	if(chrdev_buffer_size < MODULE_MIN_BUFF_SIZE){
+		printk(KERN_ERR "chrdev: chrdev_buffer is minimal size 1024 byte");
+		printk(KERN_INFO "chrdev: chrdev_buffer size is set 1024 byte");
+		chrdev_buffer_size = MODULE_MIN_BUFF_SIZE;
+	}
+		chrdev_buffer = kzalloc(chrdev_buffer_size * sizeof (*chrdev_buffer), GFP_KERNEL);
+	if (!chrdev_buffer) {
+		result = -ENOMEM;
+		printk(KERN_ERR "chrdev: chrdev_buffer is out of memory");
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/chrdev_driver.c:260]:
```suggestion
		printk(KERN_ERR "chrdev: chrdev_buffer is out of memory");
```
Think about freeing up the required resources after errors

File: 02-KernelModules_CharacterDevice/chrdev_driver.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		printk(KERN_ERR "chrdev: can't get major %d\n", chrdev_major);
+		return result;
+	}
+
+	cdev = cdev_alloc();
+ 	
+	 if (!cdev) {
+		result = -ENOMEM;
+		printk(KERN_ERR "chrdev: cdev is out of memory");
+		goto fail;
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/chrdev_driver.c]:
```suggestion
		goto fail; 
```
Think about freeing up the required resources after errors

**Review by galkinletter — CHANGES_REQUESTED** [review]
Think about freeing up the required resources after errors

**Comment by galkinletter** [comment]
Think about freeing up the required resources after errors

File: 02-KernelModules_CharacterDevice/chrdev_driver.c [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+static struct file_operations chrdev_proc = {
+        .owner = THIS_MODULE,
+        .read = proc_chrdev_read
+};
+
+
+static ssize_t chrdev_cdev_store(struct kobject *kobj, struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	int clean_flag = 0;
+    sscanf(buf, "%d", &clean_flag);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/chrdev_driver.c]: Here're issues with indentation.

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
Pl fix issues with indentation (checkpatch or whatever...)

**Comment by galkinletter** [comment]
fixed is correct

**Pull request merged**